### PR TITLE
Implement LeadingCharacterFixer

### PR DESCRIPTION
**Pull request by baile320**
This will close: https://github.com/dotenv-linter/dotenv-linter/issues/248

I have a couple comments/questions:
- When implementing this, I ran into issues with IncorrectDelimiter: it thought, for example, "*FOO=BAR" had an incorrect delimiter (this is ultimately fixed by LeadingCharacterFixer, as expected). IMO, that is not a delimiter at all, since delimiters should occur /between/ other characters, so I added a check for that in IncorrectDelimiterChecker (here: https://github.com/dotenv-linter/dotenv-linter/compare/master...baile320:leading-char-fixer?expand=1#diff-9d20ccc745eeab0a2cfbe87e10ee860cR27-R33), and removed an erroneous test.
- Depending on the leading characters, it is still possible to get erroneous UnorderedKeys warnings. I'm not sure the best way to handle this (I made no updates to that), and would like thoughts/suggestions on how to fix (if it should be fixed here) or if it's worthy of a separate PR. There's an illustrative example in my integration test (https://github.com/dotenv-linter/dotenv-linter/compare/master...baile320:leading-char-fixer?expand=1#diff-9aff8b00dfa64d0bf60b17a8b7fe7c20R4-R22)... as you can see, it throws no UnorderedKeys warnings, although it likely should, since after cleaning the keys are technically out of order (the issue is that the checking order includes the */./1 characters at the beginning of the keys, not the "cleaned" versions of the keys).

#### ✔ Checklist:
- [x] This PR has been added to [CHANGELOG.md](https://github.com/dotenv-linter/dotenv-linter/blob/master/CHANGELOG.md) (at the top of the list);
- [x] Tests for the changes have been added (for bug fixes / features);
- [ ] Docs have been added / updated (for bug fixes / features).

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 ### 🚀 Added
+- Add fixer: LeadingCharacter [#255](https://github.com/dotenv-linter/dotenv-linter/pull/259) ([@baile320](https://github.com/baile320))
```
Comment (mgrachev) [inline CHANGELOG.md:9]:
```suggestion
- Add fixer: LeadingCharacter [#259](https://github.com/dotenv-linter/dotenv-linter/pull/259) ([@baile320](https://github.com/baile320))
```

File: src/checks/incorrect_delimiter.rs [hunk 2]
```diff
@@ -126,21 +132,6 @@ mod tests {
...
-    fn leading_space_run() {
-        let mut checker = IncorrectDelimiterChecker::default();
-        let line = LineEntry {
-            number: 1,
-            file: FileEntry {
-                path: PathBuf::from(".env"),
-                file_name: ".env".to_string(),
-                total_lines: 1,
-            },
-            raw_string: String::from(" FOO=FOOBAR"),
```
Comment (mgrachev) [inline src/checks/incorrect_delimiter.rs:139]: It would be nice if you wrote a test for the case with `*FOO=BAR`.

File: src/fixes/leading_character.rs [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+    }
+}
+
+impl Fix for LeadingCharacterFixer<'_> {
+    fn name(&self) -> &str {
+        self.name
+    }
+
+    fn fix_line(&self, line: &mut LineEntry) -> Option<()> {
+        line.raw_string = line.raw_string.to_string();
```
Comment (mgrachev) [inline src/fixes/leading_character.rs:22]: I don't fully understand why you do it? Can you explain me?

**Comment by mgrachev** [comment]
> * When implementing this, I ran into issues with IncorrectDelimiter: it thought, for example, "*FOO=BAR" had an incorrect delimiter (this is ultimately fixed by LeadingCharacterFixer, as expected). IMO, that is not a delimiter at all, since delimiters should occur /between/ other characters, so I added a check for that in IncorrectDelimiterChecker (here: https://github.com/dotenv-linter/dotenv-linter/compare/master...baile320:leading-char-fixer?expand=1#diff-9d20ccc745eeab0a2cfbe87e10ee860cR27-R33), and removed an erroneous test.

Good catch! Thank you! 👍
But I'd like you to write a test for that case for the `IncorrectDelimiter`.

> * Depending on the leading characters, it is still possible to get erroneous UnorderedKeys warnings. I'm not sure the best way to handle this (I made no updates to that), and would like thoughts/suggestions on how to fix (if it should be fixed here) or if it's worthy of a separate PR. There's an illustrative example in my integration test (https://github.com/dotenv-linter/dotenv-linter/compare/master...baile320:leading-char-fixer?expand=1#diff-9aff8b00dfa64d0bf60b17a8b7fe7c20R4-R22)... as you can see, it throws no UnorderedKeys warnings, although it likely should, since after cleaning the keys are technically out of order (the issue is that the checking order includes the */./1 characters at the beginning of the keys, not the "cleaned" versions of the keys).

@evgeniy-r is already working to do it (https://github.com/dotenv-linter/dotenv-linter/issues/252)

**Comment by DDtKey** [comment]
> > * Depending on the leading characters, it is still possible to get erroneous UnorderedKeys warnings. I'm not sure the best way to handle this (I made no updates to that), and would like thoughts/suggestions on how to fix (if it should be fixed here) or if it's worthy of a separate PR. There's an illustrative example in my integration test (https://github.com/dotenv-linter/dotenv-linter/compare/master...baile320:leading-char-fixer?expand=1#diff-9aff8b00dfa64d0bf60b17a8b7fe7c20R4-R22)... as you can see, it throws no UnorderedKeys warnings, although it likely should, since after cleaning the keys are technically out of order (the issue is that the checking order includes the */./1 characters at the beginning of the keys, not the "cleaned" versions of the keys).
> 
> @evgeniy-r is already working on that feature (#252) muscle

And I think it is permissible to use the initially correct ordering in this integration test, since the target check does not order.

File: CHANGELOG.md (hunk 1 above)
Comment (baile320) [reply CHANGELOG.md]: Sorry, typo! Fixed in this commit: https://github.com/dotenv-linter/dotenv-linter/pull/259/commits/b2b50c76a30ff494b62119aed48bacf159236c04

File: src/checks/incorrect_delimiter.rs (hunk 2 above)
Comment (baile320) [reply src/checks/incorrect_delimiter.rs:139]: No problem, I fixed here: https://github.com/dotenv-linter/dotenv-linter/pull/259/commits/49d2a79a5c07495eaf14372f57dc844889e958f9

File: src/fixes/leading_character.rs (hunk 3 above)
Comment (baile320) [reply src/fixes/leading_character.rs]: That line was unnecessary, good catch, thank you :), I removed here: https://github.com/dotenv-linter/dotenv-linter/pull/259/commits/f683b815ba876f388ce3c1262d7a16b78a3aefc3 and I also extracted this function to remove the first character of the string, just in case it was unneccesarily confusing: https://github.com/dotenv-linter/dotenv-linter/pull/259/commits/3c16dc06639e0b2bb7c68c1ddd701c4dabde3bc0

File: src/checks/incorrect_delimiter.rs [hunk 4]
```diff
@@ -24,7 +24,13 @@ impl Default for IncorrectDelimiterChecker<'_> {
 impl Check for IncorrectDelimiterChecker<'_> {
     fn run(&mut self, line: &LineEntry) -> Option<Warning> {
         let key = line.get_key()?;
-        if key.trim().chars().any(|c| !c.is_alphanumeric() && c != '_') {
+
+        // delimiters occur /between/ characters, not as the initial character, so we should only check those cases
+        if key
+            .trim()
+            .starts_with(|c: char| c.is_alphabetic() || c == '_')
+            && key.trim().chars().any(|c| !c.is_alphanumeric() && c != '_')
```
Comment (DDtKey) [inline src/checks/incorrect_delimiter.rs:32]: What do you think about that case?
```
*MY-VARIABLE=test
```
it's has incorrect delimiter, but starts with `*` therefore warnings don't will be create.

Comment (baile320) [reply src/checks/incorrect_delimiter.rs:32]: Thanks for catching that. I have a fix in mind: I think we should remove **all** bad characters from the beginning of the key before we perform the IncorrectDelimiter check, does that sound reasonable?

Similarly, for the LeadingCharacterFixer, the code I have written will not turn ```***KEY=VALUE``` into ```KEY=VALUE```, it will only remove one leading character and turn it into ```**KEY=VALUE```, so I think I need to fix that that before this should be merged. I will work on that a little later today, if all of that sounds reasonable.

Comment (baile320) [reply src/checks/incorrect_delimiter.rs]: I added a helper function to ```common.rs``` called ```remove_all_invalid_leading_chars``` to do what I described above, and used it both in IncorrectDelimiterChecker and LeadingCharacterFixer. I added some additional tests that should have caught these as well.

Comment (DDtKey) [reply src/checks/incorrect_delimiter.rs]: Sorry, I have been busy last couple days.
I'll look the changes soon. 

@dotenv-linter/core what do you think? This may affect #246

File: src/checks/incorrect_delimiter.rs [hunk 5]
```diff
@@ -24,8 +24,21 @@ impl Default for IncorrectDelimiterChecker<'_> {
 impl Check for IncorrectDelimiterChecker<'_> {
     fn run(&mut self, line: &LineEntry) -> Option<Warning> {
         let key = line.get_key()?;
-        if key.trim().chars().any(|c| !c.is_alphanumeric() && c != '_') {
-            return Some(Warning::new(line.clone(), self.name(), self.message(&key)));
+
+        // delimiters occur /between/ characters, not as the initial character, so we should
+        // remove all invalid leading characters before checking for incorrect delimiters
+        let cleaned_key = remove_all_invalid_leading_chars(&key);
```
Comment (mgrachev) [inline src/checks/incorrect_delimiter.rs:30]: @baile320 Please move these changes to another PR.
We should merge these changes before merging the implemenation of `LeadingCharacterFixer`.

File: src/checks/incorrect_delimiter.rs (hunk 4 above)
Comment (mgrachev) [reply src/checks/incorrect_delimiter.rs]: I think that these changes in the `IncorrectDelimiterChecker` should be in another PR.

File: src/checks/incorrect_delimiter.rs (hunk 5 above)
Comment (baile320) [reply src/checks/incorrect_delimiter.rs:30]: I moved this logic to a separate PR: #267 . I can update this PR when #267 is merged to master, since it looks like the module file structured changed a little bit.

**Comment by baile320** [comment]
I merged in the changes from #267 and made some other relatively minor changes related to test names, checks, formatting. I think this PR is ready to go or be re-reviewed! Thanks!

File: src/fixes/leading_character.rs [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::path::PathBuf;
+
+    #[test]
+    fn fix_leading_period() {
```
Comment (mgrachev) [inline src/fixes/leading_character.rs:38]: What does `period` mean in that case?

File: src/fixes/leading_character.rs [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+                file_name: ".env".to_string(),
+                total_lines: 7,
+            },
+            raw_string: String::from("-1&*FOO=BAR"),
+        };
+
+        assert_eq!(Some(()), fixer.fix_line(&mut leading_number));
+        assert_eq!("FOO=BAR", leading_number.raw_string);
+    }
+    #[test]
```
Comment (mgrachev) [inline src/fixes/leading_character.rs:126]:
```suggestion
    }

    #[test]
```

File: src/fixes/leading_character.rs (hunk 6 above)
Comment (baile320) [reply src/fixes/leading_character.rs:38]: it just means to fix an item with a leading "." ... I have other tests for some other specific characters like a leading space, asterisk/*, etc. I can remove this test or rename if it helps to make it clearer?

Comment (mgrachev) [reply src/fixes/leading_character.rs:38]: Maybe this test should be named `fix_leading_dot`?

Comment (baile320) [reply src/fixes/leading_character.rs]: no problem, just renamed it to that! thank you

**Comment by mgrachev** [comment]
@baile320 Thanks a lot! You're awesome! ❤️
